tesseract/consensus/polygon-pos/src/notification.rs

```diff
@@ -1,101 +1,152 @@
+				let header = match header_res {
+					Ok(h) => h,
+					Err(_) => {
+						height -= 1;
+						continue;
+					},
+				};
+				let validator_set_hash_match = validate_validator_set_hash(
+					&ValidatorSet::new(trusted_state.validators.clone(), None),
+					header.header.validators_hash,
+					false,
+				);
+				let next_validator_set_hash_match = validate_validator_set_hash(
+					&ValidatorSet::new(trusted_state.next_validators.clone(), None),
+					header.header.next_validators_hash,
+					true,
+				);
+				if validator_set_hash_match.is_ok() || next_validator_set_hash_match.is_ok() {
+					found = true;
+					matched_header = Some(header);
+					break;
+				}
+				height -= 1;
+			}
+			if found {
+				let matched_height = height;
+				let matched_header = matched_header.expect("Header must be present if found");
+				let ancestry = client
+					.prover
+					.signed_headers_range(trusted_state.height + 1, matched_height)
+					.await?;
+				let next_validators = client.prover.next_validators(matched_height).await?;
+				return Ok(Some(PolygonConsensusUpdate {
+					tendermint_proof: CodecConsensusProof::from(&ConsensusProof::new(
+						matched_header,
+						ancestry,
+						Some(next_validators),
+					)),
+					milestone_update: maybe_milestone_update,
+				}));
+			}
+		},
 	}
+	Ok((None))
 }
 
-// Find the next headers to be submitted
-
-async fn next_headers(
-	prover: &PolygonPosProver,
-	chain_heads: Vec<H256>,
-	latest_block_hash: H256,
-	finalized_hash: H256,
-) -> Result<Vec<CodecHeader>, anyhow::Error> {
-	let mut headers = vec![];
-	let latest_header = prover
-		.fetch_header(latest_block_hash)
-		.await?
-		.ok_or_else(|| anyhow!("Header should exist"))?;
-	let finalized_header = prover
-		.fetch_header(finalized_hash)
+async fn build_milestone_update(
+	client: &PolygonPosHost,
+	milestone_number: u64,
+	milestone: tendermint_primitives::Milestone,
+	milestone_end_block: u64,
+) -> anyhow::Result<ismp_polygon::MilestoneUpdate> {
+	let evm_header = client
+		.prover
+		.fetch_header(milestone_end_block)
 		.await?
-		.ok_or_else(|| anyhow!("Header should exist"))?;
-	let mut parent_hash = latest_header.parent_hash;
-	headers.push(latest_header);
-	while !chain_heads.contains(&parent_hash) {
-		let header = prover
-			.fetch_header(parent_hash)
-			.await?
-			.ok_or_else(|| anyhow!("Header should exist"))?;
-		if header.number <= finalized_header.number {
-			// Unknown chain
-			headers = vec![];
-			break
-		}
-		parent_hash = header.parent_hash;
-		headers.insert(0, header)
-	}
-	Ok(headers)
+		.ok_or_else(|| anyhow::anyhow!("EVM header not found"))?;
+	let abci_query = client.prover.get_ics23_proof(milestone_number, milestone_end_block).await?;
+	let ics23_state_proof = abci_query
+		.proof
+		.as_ref()
+		.and_then(|p| p.ops.get(0))
+		.map(|op| op.data.clone())
+		.unwrap_or_default();
+	return Ok(ismp_polygon::MilestoneUpdate {
+		evm_header,
+		milestone_number: Some(milestone_number),
```

Comment:
should not be optional
```suggestion
		milestone_number: Some(milestone_number),
```